crates/cadenza-eval/src/trait_registry.rs

```diff
@@ -0,0 +1,288 @@
+
+use crate::{
+    interner::InternedString,
+    map::Map,
+    value::{MethodSignature, Type, Value},
+};
+
+/// A trait definition.
+///
+/// Traits define a set of methods that types can implement. They are similar to
+/// type classes in Haskell or traits in Rust.
+#[derive(Debug, Clone, PartialEq)]
+pub struct TraitDef {
+    /// The name of the trait.
+    pub name: InternedString,
+    /// The method signatures defined by this trait.
+    pub methods: Vec<MethodSignature>,
+    /// Type parameters for generic traits (currently unused, reserved for future).
+    pub type_params: Vec<InternedString>,
+}
+
+impl TraitDef {
+    /// Creates a new trait definition.
+    pub fn new(name: InternedString, methods: Vec<MethodSignature>) -> Self {
+        Self {
+            name,
+            methods,
+            type_params: Vec::new(),
+        }
+    }
+
+    /// Creates a new trait definition with type parameters.
+    pub fn with_type_params(
+        name: InternedString,
+        methods: Vec<MethodSignature>,
+        type_params: Vec<InternedString>,
+    ) -> Self {
+        Self {
+            name,
+            methods,
+            type_params,
+        }
+    }
+}
+
+/// A trait implementation.
+///
+/// Connects a type to a trait and provides the method implementations.
+#[derive(Debug, Clone, PartialEq)]
+pub struct TraitImpl {
+    /// The name of the trait being implemented.
+    pub trait_name: InternedString,
+    /// The type for which the trait is being implemented.
+    pub for_type: Type,
+    /// The method implementations.
+    /// Maps method name to the implementation (a function value).
+    pub methods: Map<Value>,
+}
+
+impl TraitImpl {
+    /// Creates a new trait implementation.
+    pub fn new(trait_name: InternedString, for_type: Type, methods: Map<Value>) -> Self {
+        Self {
+            trait_name,
+            for_type,
+            methods,
+        }
+    }
+}
+
+/// A registry for trait definitions and implementations.
+///
+/// The trait registry stores all trait definitions and implementations in the program.
+/// It provides lookup methods to find traits and their implementations for specific types.
+pub struct TraitRegistry {
+    /// All defined traits, indexed by trait name.
+    traits: Map<TraitDef>,
+    /// All trait implementations.
+    /// Key: (type, trait name) -> implementation
+    /// Using a Vec instead of a Map because the key is composite
+    implementations: Vec<TraitImpl>,
+}
+
+impl TraitRegistry {
+    /// Creates a new empty trait registry.
+    pub fn new() -> Self {
+        Self {
+            traits: Map::default(),
+            implementations: Vec::new(),
+        }
+    }
+
+    /// Registers a new trait definition.
+    ///
+    /// # Errors
+    /// Returns an error if a trait with the same name already exists.
+    pub fn define_trait(&mut self, trait_def: TraitDef) -> Result<(), String> {
```

Comment:
We should always be returning diagnostics for errors so we can attach the message to spans